src/psyclone/psyir/symbols/datasymbol.py

```diff
@@ -277,19 +278,28 @@ def __str__(self):
             ret += ", constant_value={0}".format(self.constant_value)
         return ret + ">"
 
-    def copy(self):
+    def copy(self, interface=None, visibility=None):
```

Comment:
I do understand adding a `visibility=self.visibility` that is missing inside the copy.

Reply:
I agree that it does seem over engineered. I ended up doing this because the `visibility` of a Symbol is currently immutable and therefore it needs to be supplied to the constructor (called from within `copy`). I then ended up with code duplication and therefore took this approach to remove it - the common code is now in `_base_copy()`. It was complicated because the constructors of DataSymbol and Symbol have different, required arguments. I think we have three choices:
1. allow the visibility to be mutable;
2. we live with the duplication of the 'do we override the visibility?' code in each of the two copy routines;
3. what I currently have.

(The duplication in 2. is less then I first through because we could just set the new interface after the call to copy(), as you suggest.) I think my preferred option is 2. but I'm open to persuasion :-)

Reply:
I think number 1) would be the easier. Is there any reason for Visibility to be immutable? If an external code links to a symbol and we change its visibility the compilation will fail at some point, but this seems an user error not necessarily something psyclone should prevent.

I think my problem is that then we introduced this method I was thinking on something akin to the OO copy constructor and we are now giving it specific functionalities that go beyond the simple copy semantics.

It also allows the implementation to be simpler:
generic copy be: `return type(self)(self.name, self.visibility, self.interface)`
DataSymbol copy be: `return DataSymbol(self.name, self.datatype, self.visibility, self.constant_value, self.interface)`
but this is a secondary consideration.

I am also open to be persuaded of the contrary.

Reply:
That's true, 1. is easiest. I was assuming that you wanted it to keep it immutable!